src/ImageSharp.Web/Synchronization/RefCountedConcurrentDictionary.cs

```diff
@@ -0,0 +1,246 @@
+                    {
+                        this.valueReleaser?.Invoke(refCountedValue.Value);
+                        return;
+                    }
+                }
+                else
+                {
+                    // Decrement ref count
+                    if (this.dictionary.TryUpdate(key, new RefCountedValue(refCountedValue.Value, refCountedValue.RefCount - 1), refCountedValue))
+                    {
+                        return;
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get an enumeration over the contents of the dictionary for testing/debugging purposes
+        /// </summary>
+        internal IEnumerable<(TKey Key, TValue Value, int RefCount)> DebugGetContents() => new RefCountedDictionaryEnumerable(this);
+
+        /// <summary>
+        /// Internal class used for testing/debugging purposes
+        /// </summary>
+        private class RefCountedDictionaryEnumerable : IEnumerable<(TKey Key, TValue Value, int RefCount)>
+        {
+            private readonly RefCountedConcurrentDictionary<TKey, TValue> dictionary;
+
+            internal RefCountedDictionaryEnumerable(RefCountedConcurrentDictionary<TKey, TValue> dictionary)
+                => this.dictionary = dictionary;
+
+            public IEnumerator<(TKey Key, TValue Value, int RefCount)> GetEnumerator()
+                => new RefCountedDictionaryEnumerator(this.dictionary);
+
+            IEnumerator IEnumerable.GetEnumerator() => this.GetEnumerator();
+        }
+
+        /// <summary>
+        /// Internal class used for testing/debugging purposes
+        /// </summary>
+        private class RefCountedDictionaryEnumerator : IEnumerator<(TKey Key, TValue Value, int RefCount)>
+        {
+            private readonly IEnumerator<KeyValuePair<TKey, RefCountedValue>> enumerator;
+
+            public RefCountedDictionaryEnumerator(RefCountedConcurrentDictionary<TKey, TValue> dictionary)
+                => this.enumerator = dictionary.dictionary.GetEnumerator();
+
+            public (TKey Key, TValue Value, int RefCount) Current
+            {
+                get
+                {
+                    KeyValuePair<TKey, RefCountedValue> keyValuePair = this.enumerator.Current;
+                    return (keyValuePair.Key, keyValuePair.Value.Value, keyValuePair.Value.RefCount);
+                }
+            }
+
+            object IEnumerator.Current => this.Current;
+
+            public void Dispose() => this.enumerator.Dispose();
+
+            public bool MoveNext() => this.enumerator.MoveNext();
+
+            public void Reset() => this.enumerator.Reset();
+        }
+
+        /// <summary>
+        /// Simple immutable tuple that combines a <typeparamref name="TValue"/> instance with a ref count integer.
+        /// </summary>
+        private class RefCountedValue : IEquatable<RefCountedValue>
```

Comment:
Can you use struct records here instead ?

Reply:
I benchmarked this using both class and struct for this type and determined that using a class executes more quickly and allocates less memory.  I believe this happens because `ConcurrentDictionary` can use an optimized code path that leverages atomic writes when TValue is a class, but has to fall back on a less efficient path that allocates when TValue is a struct.
```
|                          Method |      Mean |    Error |   StdDev |  Gen 0 | Allocated |
|-------------------------------- |----------:|---------:|---------:|-------:|----------:|
|       Class_GetAndReleaseNewKey | 129.92 ns | 0.554 ns | 0.518 ns | 0.0095 |      80 B |
|      Struct_GetAndReleaseNewKey | 147.59 ns | 0.687 ns | 0.573 ns | 0.0067 |      56 B |
|  Class_GetAndReleaseExistingKey | 142.32 ns | 1.159 ns | 1.027 ns | 0.0076 |      64 B |
| Struct_GetAndReleaseExistingKey | 177.69 ns | 0.813 ns | 0.721 ns | 0.0134 |     112 B |
|            Class_GetExistingKey |  69.75 ns | 0.301 ns | 0.267 ns | 0.0038 |      32 B |
|           Struct_GetExistingKey |  89.81 ns | 0.682 ns | 0.638 ns | 0.0067 |      56 B |
```

I have not yet experimented with using record types (in part because this codebase is still targeting netcoreapp2.1 and 3.1).

Reply:
Let's leave changing target frameworks to V2. I don't want your hard work delayed by our upstream work.

Reply:
If not a struct record, tuples would still be good while keeping the code simpler. But I don't know about tfm requirements either, so you'll decide.